File: src/0235.lowest-common-ancestor-of-a-binary-search-tree.235/solution.go
package lowestcommonancestorofabinarysearchtree235

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if p.Val > q.Val {
		tmp := p
		p = q
		q = tmp
	}

	return find(root, p, q)
}

func find(root, p, q *TreeNode) *TreeNode {
	if root == p || root == q {
		return root
	}
	if root.Val > p.Val && root.Val < q.Val {
		return root
	}
	if root.Val < p.Val {
		return find(root.Right, p, q)
	} else {
		return find(root.Left, p, q)
	}

}

File: src/0235.lowest-common-ancestor-of-a-binary-search-tree.235/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (p.val < q.val) {
            return search(root, p, q);
        }
        return search(root, q, p);
    }

    public TreeNode search(TreeNode root, TreeNode p, TreeNode q) {
        if (root.val == p.val || root.val == q.val) {
            return root;
        }
        if (root.val > p.val && root.val < q.val) {
            return root;
        }
        if (root.val > q.val) {
            return search(root.left, p, q);
        }
        return search(root.right, p, q);
    }

}